pisa/stages/flux/pi_mceq_barr_red.py

```diff
@@ -0,0 +1,492 @@
+    then multiplied with a shift in spectral index, and then modifications due
+    to meson production (barr variables) are added.
+
+    The MCEq-table has 2 solutions of the cascade equation per Barr variable (12)
+    - one solution for meson and one solution for the antimeson production uncertainty.
+
+    Each solution consists of 4 splines: "dnumu", "dnumubar", "dnue", and
+    "dnuebar". These are the gradients that govern how the neutrino fluxes vary
+    depending on modifications to the Barr params.
+
+    """
+
+    def __init__(
+        self,
+        table_file,
+        data=None,
+        params=None,
+        input_names=None,
+        output_names=None,
+        debug_mode=None,
+        input_specs=None,
+        calc_specs=None,
+        output_specs=None,
+    ):
+
+        #
+        # Define parameterisation
+        #
+
+        # Define the Barr parameters
+        # TODO Barr block definition could change - need to make this more flexible
+        # Perhaps let user define the blocks, and then table is generated on the
+        # fly as a first step? Could take up to 1 hour to produce table though...
+        self.barr_param_names = [
+            # pions
+            "af",
+            "g",
+            "h",
+            "i",
+            # kaons
+            "w",
+            "x",
+            "y",
+            "z",
+        ]
+
+        # Define signs for Barr params
+        # +  -> meson production
+        # -  -> antimeson production
+        self.barr_param_signs = ["+", "-"]
+
+        # Atmopshere model params
+        # TODO
+
+        # Get the overall list of params for which we have gradients stored
+        # Define a mapping to index values, will be useful later
+        self.gradient_param_names = [
+            n + s for n in self.barr_param_names for s in self.barr_param_signs
+        ]
+        self.gradient_param_indices = collections.OrderedDict(
+            [(n, i) for i, n in enumerate(self.gradient_param_names)]
+        )
+
+        #
+        # Call stage base class constructor
+        #
+
+        # Define stage parameters
+        expected_params = (
+            # pion
+            "pion_ratio",
+            "barr_af_Pi",
+            "barr_g_Pi",
+            "barr_h_Pi",
+            "barr_i_Pi",
+            # kaon
+            "barr_w_K",
+            "barr_x_K",
+            "barr_y_K",
+            "barr_z_K",
+            "barr_w_antiK",
+            "barr_x_antiK",
+            "barr_y_antiK",
+            "barr_z_antiK",
+            # CR
+            "delta_index",
+            "energy_pivot",
+        )
+
+        input_names = ()
+        output_names = ()
+
+        # Using Honda for nominal flux. Keys should already exist
+        input_calc_keys = ("nu_flux_nominal", "nubar_flux_nominal")
+        # what are keys added or altered in the calculation used during apply
+        output_calc_keys = ("nu_flux")
```

Comment:
Ah I know! These are not tuples but strings. It should be `("nu_flux", )` here and below

Reply:
excellent ok, and then what will be the three lines of code to link containers? Is there an example?

Reply:
Cool! Does that still produce sensible flux systematics?
Linking containers is for example used here: https://github.com/IceCubeOpenSource/pisa/blob/master/pisa/stages/flux/pi_honda_ip.py#L84

In your case, I think the calculation only depends weather it's nu or anti-nu, so you would link all nu containers together and all anti-nu. Do that *only* when cal_mode == 'binned', and after iterating though the containers unlink those again....that should be all. That way the calculation is really only done twice instead of 12 times

Reply:
- Calculating in binned mode now -
 
Before linking containers: 
![binned_no_links](https://user-images.githubusercontent.com/30318289/79984906-1cc1ef80-84aa-11ea-8ebd-53a7a033206b.png)

After linking containers 
![binned_with_links](https://user-images.githubusercontent.com/30318289/79984909-1d5a8600-84aa-11ea-9aca-f027e7a603c4.png)

Looks the same, was definitely MUCH faster! Alles gut :)